Add guidance on PAT Auth for Private Packages

ben-lavelle opened the pull request:
I'm not really sure where is the right place for this, but after a long struggle getting a private package to install, @samtazzyman and I thought it worth writing this info up somewhere. Apologies for the delay, it was only when a colleague came to me with the `404` problem that I finally was pushed into action--looks like package development is gently starting to take off!

🔵 **New Content:**
- Preamble on Public / Private repos in the organisation (couldn't see this anywhere else--this might not be the best place for it / might be repeated elsewhere)
- Explanation of `404` errors installing Private packages using `remotes::`/`devtools::` in R.
- Walkthrough getting a PA, saving it to system environment, and using it in conjunction with `remotes::install_github`.
- Some narrative on security considerations--these aren't exhaustive, it's only my (very uninformed) view of it from muddling around.

⚠️ **Known Issues:**
- 🔴 **I haven't been able to compile this** (sorry) so i don't know if it renders; it was all edited on GitHub so doesn't preview. I did briefly clone and try to `bookdown` it on R but threw a bunch of errors I haven't had a chance to shoot yet 🐑 
- Environment variables like `GITHUB_PAT` don't seem to persist like I think they ought to on a local machine running R. `.RProfile`s also don't seem to persist, and don't seem to run on every startup (I presume because R never really 'shuts down' in the Docker universe in the same way?). I think this means we have a problem with not having any cross-project persistent storage things like this? 🌵 

Feel free to move around/clean up/wipe if it's unuseful. Cheers!

ben-lavelle commented:
Can I bump this for review? Had someone else come to me asking about it today.

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -192,6 +192,62 @@ The platform has configured simple "safety barriers" to reduce risk of accidenta
+
+### Private and Public Repositories
+
+The majority of repos in the DASD GitHub organisation are **Private**: this keeps visibility, access and editing rights controlled to particular groups within the MoJ network and is sensible in most cases. Owners and Admins of a repository can control this visibility by going to **_Settings_ > _Collaborators & Teams_** and adding teams to read/write/admin access groups. To maximise collaboration  ...
```
Yeah I am uneasy about telling people to code in private - I agree that we should generally be coding in the open (although my own repos are typically private).

However I appreciate that there is a cultural resistance to sharing things that we don't yet feel are 'ready'.

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -192,6 +192,62 @@ The platform has configured simple "safety barriers" to reduce risk of accidenta
+
+For the most part, this private group-based access works well and seamlessly. 
+Where it can rub is in building **R Packages** for high-quality reproducable analysis either across small teams or across the organisation.
+The AP team have built a number of _generic_ tools that are non-sensitive in nature (e.g. [`s3tools`](github.com/moj-analytical-services/s3tools) and [`dbtools`](github.com/moj-analytical-services/dbtools)) and as such are in fact **Public** repos that anyone, inside or outside the organisation, can view, access, and install to their library.
```
Agree

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -192,6 +192,62 @@ The platform has configured simple "safety barriers" to reduce risk of accidenta
+For the most part, this private group-based access works well and seamlessly. 
+Where it can rub is in building **R Packages** for high-quality reproducable analysis either across small teams or across the organisation.
+The AP team have built a number of _generic_ tools that are non-sensitive in nature (e.g. [`s3tools`](github.com/moj-analytical-services/s3tools) and [`dbtools`](github.com/moj-analytical-services/dbtools)) and as such are in fact **Public** repos that anyone, inside or outside the organisation, can view, access, and install to their library.
+If an internal package is being used for a specific project, RAP or other analytical work, the chances are that it will not be appropriate to make these codebases public, as some combination of the structure, code and comments can be revealing, however [sensibly the data are stored](https://moj-analytical-services.github.io/platform_user_guidance/working-with-big-datasets-and-sharing-them-accessi ...
```
Definitely agree with this, particularly for packages for general RAP functions, because they shouldn't contain information about the data to which they are being applied.

davidread commented:
Perhaps now @ben-lavelle has done the hard work getting all this down originally, @calumabarnett you could help move this on by doing some editing, taking into account the comments and more thoughts from @ben-lavelle?

ben-lavelle commented:
> Perhaps now @ben-lavelle has done the hard work getting all this down originally, @calumabarnett you could help move this on by doing some editing, taking into account the comments and more thoughts from @ben-lavelle?

I've got it down on my Friday jobs to review and make some changes; thanks all for very thoughtful input, it raised more fundamental questions than I expected! Happy to fix small things before Calum has a look at bigger issues to save him some time.

davidread commented:
Perfect, thanks @ben-lavelle!

calumabarnett commented:
@ben-lavelle I think the main thing for now is to make sure the content is accurate and correctly reflects our position. I'm in the process of updating the guidance as a whole and may want to rewrite and restructure some things – it might make most sense to merge this section once you've reviewed the comments and I can get your review on any edits or changes I make when I get to this section.

davidread commented:
I've made some edits about the private/public repo stuff. Hopefully this is a reasonable compromise.

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -196,19 +196,23 @@ The platform has configured simple "safety barriers" to reduce risk of accidenta
+GitHub repositories can be public or private:
 
-### Private R Packages for Reproducible Analysis
+*Public repos* means visible to the world. A repo should contain no passwords/secrets and no data (apart from small reference tables) - these are particulary important for public repos.
```
I think this bit about
>"A repo should contain no passwords/secrets and no data (apart from small reference tables)"

ought to be above the *Public repos* section, immediately following the statement that Github repos can be public or private. We can then add the particular importance of this for public repos here to really drive it home.

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -220,35 +224,41 @@ A Personal Access Token (PAT) is provided by GitHub through each user's profile,
-  3. Copy your PAT into a `.txt` file that you can keep somewhere securely. You won't be able to come back to this page to see this PAT again, and will have to generate a new one if you lose track of it.
+  3. Copy your PAT to the clipboard, for immediate use - see the next section.
+  
+     Note: GitHub won't show this PAT to you ever again BUT you shouldn't save it in a file or anything. If you were to have passwords in plain text lying around your hard disk then it becomes a security hazard. If you need a PAT in future (eg you wiped your .simply create a new one, and invalidate the previous one.
```
I think we can legitimately strengthen the tone of this, e.g. "don't save it in a file or anything" rather than "you shouldn't". Also "Passwords in plain text saved in a file anywhere on your hard disk are a security hazard" rather than "If you were to have passwords in plain text lying around your hard disk then it becomes a security hazard". (No need to follow my exact wording)

Also there's a typo: "(eg you wiped your .simply create..."

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -220,35 +224,41 @@ A Personal Access Token (PAT) is provided by GitHub through each user's profile,
-  ### Using a PAT to Authenticate in R/RStudio
+### Using a PAT to Authenticate in R/RStudio
 
 A safe, fast, and reliable method is to store the PAT in the so-called **system environment**.
```
Is there an obvious reason as to why storing the PAT in the system environment is less of a security risk than storing on your own local machine? In case people ask, rather than needing to reflect it here.

davidread replied:
The main thing we're fighting here is people committing the PAT into the repo. The reason is the 'need to know' principle. We've now mentioned that the PAT needs to be kept private to the user, and linked to the MoJ security guidance which talks more about this.

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -220,35 +224,41 @@ A Personal Access Token (PAT) is provided by GitHub through each user's profile,
 A safe, fast, and reliable method is to store the PAT in the so-called **system environment**.
 This is an area where basic, 'long-term' variables are kept, like your Operating System details, that do not change, even as you switch between analyses, sessions or projects.
-Conveniently, most [`r-lib`-style](https://github.com/r-lib) packages (those built by the inimitable Hadley Wickham and friends--for most people a backbone of R infrastructure) including `remotes` and `devtools`, will look for a PAT in the system environment by default, under the name `GITHUB_PAT`, so no modifications need be made to any project code once a user has registered a PAT in this locat ...
+Conveniently, most [`r-lib`-style](https://github.com/r-lib) packages (those built by Hadley Wickham and friends--for most people a backbone of R infrastructure) including `remotes` and `devtools`, will look for a PAT in the system environment by default, under the name `GITHUB_PAT`, so no modifications need be made to any project code once a user has registered a PAT in this location.
```
I don't know that we even need to namecheck Hadley Wickham, do we? Can we just say something like "those that are part of the tidyverse, for many people a backbone of R infrastructure"?

(☝️ assuming what I've written about the Tidyverse is actually true)

ben-lavelle replied:
My only intention here was that I'm not sure people know what the tidyverse is but often people are directed to (or at least I direct people to) Hadley's 'R for Data Science' so they might better relate it to their work. I promise it's not a personal obsession with HW, promise...

davidread replied:
I've trimmed out the tidyverse stuff as it is a bit off track. It now says `[`GITHUB_PAT` env variable] is where common R packages (e.g. remotes and devtools) will look for it.`

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -220,35 +224,41 @@ A Personal Access Token (PAT) is provided by GitHub through each user's profile,
 This is an area where basic, 'long-term' variables are kept, like your Operating System details, that do not change, even as you switch between analyses, sessions or projects.
-Conveniently, most [`r-lib`-style](https://github.com/r-lib) packages (those built by the inimitable Hadley Wickham and friends--for most people a backbone of R infrastructure) including `remotes` and `devtools`, will look for a PAT in the system environment by default, under the name `GITHUB_PAT`, so no modifications need be made to any project code once a user has registered a PAT in this locat ...
+Conveniently, most [`r-lib`-style](https://github.com/r-lib) packages (those built by Hadley Wickham and friends--for most people a backbone of R infrastructure) including `remotes` and `devtools`, will look for a PAT in the system environment by default, under the name `GITHUB_PAT`, so no modifications need be made to any project code once a user has registered a PAT in this location.
 This is the approach we'll make use of:
```
Probably change to something like "This is the recommended approach:"

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
````diff
@@ -220,35 +224,41 @@ A Personal Access Token (PAT) is provided by GitHub through each user's profile,
-  ```
   
+      ```{r, eval = FALSE}
+      Sys.setenv("GITHUB_PAT" = "ax451...8838b1")
````
Do we need to be clear that these lines of code should under no circumstances end up in the repo? I realise that's repetitive but as it stands I could see people mistakenly pasting the above snippet into their `global.R` file or whatever, and then `git push origin` and boom. At the very least if they think that's what we mean there's a danger of seeming self-contradictory.

davidread replied:
I've changed this line to be a bit of shell script that writes the correct R file, and I've added another warning for luck.

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
````diff
@@ -220,35 +224,41 @@ A Personal Access Token (PAT) is provided by GitHub through each user's profile,
-  ```
+
+      ```{r, eval = FALSE}
+      remotes::install_github("moj-analytical-services/[private-package]@v1.6")
````
so cool

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -220,35 +224,41 @@ A Personal Access Token (PAT) is provided by GitHub through each user's profile,
-Because of the pod system on which the Platform works, system environment variables are _not_ always as persisent as hoped. It's worth putting the `setenv` call above (1) into an `.RProfile` or other file **that you do not upload to GitHub** to give easy, but private access. Keep a copy of your PAT on your local machine. You can use `Sys.getenv("GITHUB_PAT")` to check whether a PAT is loaded.
 
-## Other tips and tricks [Work in progress!]:
+Because of the pod system on which the Platform works, system environment variables are _not_ always as persistent as hoped. It's worth putting the `setenv` call above (1) into an `.RProfile` or other file **that you do not upload to GitHub** to give easy, but private access. Keep a copy of your PAT on your local machine. You can use `Sys.getenv("GITHUB_PAT")` to check whether a PAT is loaded.
```
Hang on - didn't we just tell them _not_ to keep a copy of the PAT on their local machine?

Also do we really want them to put their `setenv` call into `.Rprofile`? This seems risky because it could easily be accidentally pulled into github. Might it be more secure to ask them to create an `Renv` file or something similar that they can put their `setenv` files in, and then immediately add this to their `.gitignore` to prevent a visit from Mr Cockup?

davidread replied:
This is now addressed.

ben-lavelle commented:
Hi all, sorry for the sluggishness, I am re-engaging with this and will go back to the details above, but just to go in for the central point about saving or not-saving this PAT, I did a quick check just now and I think my understanding was a bit off before--it really **definitely doesn't save between projects/sessions** if you just run the `Sys.setenv()` once. If I set it then change project or even just Ctrl+Shift+F10 it's gone--based on the above any user would then have to regenerate and reset a PAT every session they needed it (though granted it's only necessary for installation so it wouldn't be every day for most, but it'd be at least every release of any package).

However, I was looking at the reading on `.Rprofile` and the logic is that R looks **on startup** for *"a `.Rprofile` file in the current [project] directory, and then if there isn't one it looks in the user's home directory"*. I trialled creating a one-line `.Rprofile` with the `Sys.setenv` call in my home user directory, and it

- Remembered my PAT when I restarted the session
- Remembered my PAT when I changed to a project without its own profile
- Forgot my PAT when I removed the profile and restarted the session
- Forgot my PAT when I changed to a project *with* a competing profile
- Changed my PAT when I changed the file and restarted the session

i.e. behaved exactly as hoped.

I'd then propose this home-directory `.Rprofile` as by far the most functional solution, and I'd argue security-wise that this makes it *most* likely that people wouldn't copy their PAT into GitHub code (as the home directory is never tracked, and the PAT lives long enough that most would forget they'd ever set it) so the only flaw would be someone being able to access the user's analytical platform user area *and* know what they were looking for. Even in this case, we can strengthen guidance on the scope of the PAT such that all the PAT authorises is read-only access to private repos. If an attacker had got that far they'd already have surpassed this level of access.

ben-lavelle commented:
Have just tidied up little comments where I think there's agreement; saved the big hitters for later.

davidread commented:
I agree @ben-lavelle . I've now changed the guidance to create the home-directory .Rprofile file. If you wouldn't mind checking this is ok?

Security-wise I think it is fine to have the PAT in this file because it only has 'repo read' permissions - nothing more powerful. I've added a comment that if you want a PAT with any more scope then take advice about how to secure the PAT.

davidread commented on 04-using-github-with-the-platform.Rmd:
```diff
@@ -24,12 +24,11 @@ Once signed in, go to the MoJ Analytical Services homepage at https://github.com
 
-  ![](images/github/create_repo.gif)
-
+![create repo](images/github/create_repo.gif)
```
I've added some alt text like this for accessibility reasons.

ben-lavelle commented on 04-using-github-with-the-platform.Rmd:
````diff
@@ -192,7 +193,89 @@ The platform has configured simple "safety barriers" to reduce risk of accidenta
+Set this up by running this in your linux terminal (ie R Studio's 'terminal') and substitute your own PAT Token:
+  
+```{bash}
+echo 'Sys.setenv("GITHUB_PAT" = "ax451...8838b1")' >> ~/.Rprofile
````
Note that users can't Ctrl+V into the terminal, you just return the keypress ``^V`. Have to Ctrl+Alt+V or right click and paste.

davidread replied:
How about some single quotes?
```
writeLines('Sys.setenv("GITHUB_PAT" = "ax451...8838b1"', con = "~/.Rprofile")
```

ben-lavelle replied:
Neat. Popped in a missing bracket.

samtazzyman commented on 04-using-github-with-the-platform.Rmd:
````diff
@@ -192,7 +193,89 @@ The platform has configured simple "safety barriers" to reduce risk of accidenta
+writeLines('Sys.setenv("GITHUB_PAT" = "ax451...8838b1")', con = "~/.Rprofile")
+```
+
+**Warning: This line should not be put in any repo. And this .Rprofile file should not be added to any repo.**
````
Can we put this warning immediately before the line in question, rather than (or possibly as well as) immediately after? I worry about people greedily copying and pasting before they read on.

ben-lavelle replied:
Yep, in.
